pkg/entrypoint/entrypointer.go

```diff
@@ -467,3 +501,189 @@ func (e Entrypointer) outputRunResult(terminationReason string) result.RunResult
+//
+// The artifact template is expected to be in the format "$(steps.{step-name}.outputs.{artifact-name})" or "$(steps.{step-name}.outputs)".
+func parseArtifactTemplate(template string) (ArtifactTemplate, error) {
+	if template == "" {
+		return ArtifactTemplate{}, errors.New("template is empty")
+	}
+	if artifactref.StepArtifactRegex.FindString(template) != template {
+		return ArtifactTemplate{}, fmt.Errorf("invalid artifact template %s", template)
+	}
+	template = strings.TrimSuffix(strings.TrimPrefix(template, "$("), ")")
+	split := strings.Split(template, ".")
+	at := ArtifactTemplate{
+		ContainerName: "step-" + split[1],
+		Type:          split[2],
+	}
+	if len(split) == 4 {
+		at.ArtifactName = split[3]
+	}
+	return at, nil
+}
+
+// ArtifactTemplate holds steps artifacts metadata parsed from step artifacts interpolation
+type ArtifactTemplate struct {
+	ContainerName string
+	Type          string // inputs or outputs
+	ArtifactName  string
+}
+
+// applyStepArtifactSubstitutions replaces artifact references within a step's command and environment variables with their corresponding values.
+//
+// This function is designed to handle artifact substitutions in a script file, inline command, or environment variables.
+//
+// Args:
+//
+//	stepDir: The directory of the executing step.
+//
+// Returns:
+//
+//	An error object if any issues occur during substitution.
+func (e *Entrypointer) applyStepArtifactSubstitutions(stepDir string) error {
+	// Script was re-written into a file, we need to read the file to and substitute the content
+	// and re-write the command.
+	// While param substitution cannot be used in Script from StepAction, allowing artifact substitution doesn't seem bad as
+	// artifacts are unmarshalled, should be safe.
+	if len(e.Command) == 1 && filepath.Dir(e.Command[0]) == filepath.Clean(ScriptDir) {
+		dataBytes, err := os.ReadFile(e.Command[0])
+		if err != nil {
+			return err
+		}
+		fileContent := string(dataBytes)
+		v, err := replaceValue(artifactref.StepArtifactRegex, fileContent, stepDir, getArtifactValues)
+		if err != nil {
+			return err
+		}
+		if v != fileContent {
+			temp, err := writeToTempFile(v)
+			if err != nil {
+				return err
+			}
+			e.Command = []string{temp.Name()}
+		}
+	} else {
+		command := e.Command
+		var newCmd []string
+		for _, c := range command {
+			v, err := replaceValue(artifactref.StepArtifactRegex, c, stepDir, getArtifactValues)
+			if err != nil {
+				return err
+			}
+			newCmd = append(newCmd, v)
+		}
+		e.Command = newCmd
+	}
+
+	// substitute env
+	for _, e := range os.Environ() {
+		pair := strings.SplitN(e, "=", 2)
+		v, err := replaceValue(artifactref.StepArtifactRegex, pair[1], stepDir, getArtifactValues)
+
+		if err != nil {
+			return err
+		}
+		os.Setenv(pair[0], v)
+	}
+
+	return nil
+}
+
+func writeToTempFile(v string) (*os.File, error) {
+	tmp, err := os.CreateTemp("", "script-*")
```

Comment:
For some reason, this does not work for me.

This step:
```golang
      - name: artifacts-consumer-python
        image: python:latest
        script: |
          #!/usr/bin/env python3
          import json

          data = json.loads('$(steps.artifacts-producer.outputs.image)')
          print(f'URL: {data["url"]}, Digest: {data["digest"]}')
```

Gives me:

```shell
[artifacts-consumer-inputs] 2024/04/11 12:30:37 Error executing command: fork/exec /tmp/script-1761841485: no such file or directory

[artifacts-consumer-python] 2024/04/11 12:30:38 Skipping step because a previous step failed

container step-artifacts-consumer-python has failed  : [{"key":"StartedAt","value":"2024-04-11T12:30:38.131Z","type":3},{"key":"Reason","value":"Skipped","type":3}]
```

Reply:
Never mind, the error was on my side, in a previous step:

```yaml
      - name: artifacts-consumer
        image: bash:latest
        script: |
          echo $(steps.artifacts-producer.outputs)
      - name: artifacts-consumer-images
        image: bash:latest
        script: |
          echo $(steps.artifacts-producer.outputs.image)
      - name: artifacts-consumer-inputs
        image: bash:latest
        script: |
          #!/bin/bash
          echo $(steps.artifacts-producer.inputs.source)
      - name: artifacts-consumer-python
        image: python:latest
        script: |
          #!/usr/bin/env python3
          import json

          data = json.loads('$(steps.artifacts-producer.outputs.image)')
          print(f'URI: {data[0]["uri"]}, Digest: {data[0]["digest"]}')
```

The problem was `#!/bin/bash` but the error message was confusing.
Once I replace that with `#!/bin/sh` or `#!/usr/bin/env bash` it works fine.